docs/content/docs/upgrade_guide.md

```diff
@@ -0,0 +1,101 @@
+- Breaking changes may be introduced between `0.y.z` and `0.y+1.0`
+
+For OSM versions `x.y.z` where `x >= 1`:
+- Breaking changes will not be introduced between `x.y.z` and `x.y+1.0` or between `x.y.z` and `x.y.z+1`
+- Breaking changes may be introduced between `x.y.z` and `x+1.0.0`
+
+## Resource availability during upgrade
+Since upgrades include redeploying the osm-controller with the new version, there may be some downtime of the controller. This means that there will be a delay in processing new SMI resources. 
+
+However, already existing SMI resources will be unaffected (assuming [CRD Upgrades](#CRD-Upgrades) are not needed). This means that the data plane (which includes the Envoy sidecar configs) will also be unaffected by upgrading. 
+
+Data plane interruptions are expected if the upgrade includes CRD changes. Streamlining data plane upgrades is being tracked in issue [#512](https://github.com/openservicemesh/osm/issues/512).
+
+## How to upgrade OSM
+
+The recommended way to upgrade a mesh is with the `osm` CLI. For advanced use cases, `helm` may be used.
+
+### CRD Upgrades
+Because Helm does not manage CRDs beyond the initial installation, special care needs to be taken during upgrades when CRDs are changed. Please check the `CRD Updates` section of the [release notes](https://github.com/openservicemesh/osm/releases) to see if additional steps are required to update the CRDs used by OSM. If the new release does contain updates to the CRDs, it is required to first delete existing CRDs and the associated Custom Resources prior to upgrading.
+
+In the `./scripts/cleanup` directory we have included a helper script to delete those CRDs and Custom Resources: `./scripts/cleanup/crd-cleanup.sh`
+
+After upgrading, the CRDs and Custom Resources will need to be recreated.
+
+1. Checkout the tag of the [repo](https://github.com/openservicemesh/osm) corresponding to the version of the upgraded chart.
+1. Install the new CRDs. (Run from the root of the repo.)
+    - `kubectl apply -f charts/osm/crds/`
+1. Recreate CustomResources
+
+### Upgrading with the OSM CLI
+
+**Pre-requisites**
+
+- Kubernetes cluster with the OSM control plane installed
+- `osm` CLI installed
+  - By default, the `osm` CLI will upgrade to the same chart version that it installs. e.g. v0.8.0 of the `osm` CLI will upgrade to v0.8.0 of the OSM Helm chart.
+
+The `osm mesh upgrade` command performs a `helm upgrade` of the existing Helm release for a mesh.
+
+Basic usage requires no additional arguments or flags:
+```console
+$ osm mesh upgrade
+```
+
+This command will upgrade the mesh with the default mesh name in the default OSM namespace. Values from the previous release will carry over to the new release except for `OpenServiceMesh.image.registry` and `OpenServiceMesh.image.tag` which are overridden by default. For example, if OSM v0.7.0 is installed, `osm mesh upgrade` for v0.8.0 of the CLI will update the control plane images to v0.8.0 by default.
```

Comment:
I see three main options to handle this scenario:
1. We add some conversion logic of the old values to the new values to `osm mesh upgrade`. This would be automatic and transparent, but could turn into a pain to maintain and further couples the CLI with the chart.
2. We document the mapping and let users do it themselves with `helm upgrade`.
3. We ignore any remapping, declare the change breaking, and tell users to do a clean install.

Either way, what changes could we make now before we have a need for them?

Reply:
Yea each of these has pros/cons. I think option 1 makes the most sense to me. The CLI is in fact dependent on the helm chart as it is a core component of the install/lifecycle management of the control plane, so embedding the logic to remap values there makes sense to me. Agree it might be a pain to manage but since we only support one minor version upgrade at the moment, we can maintain only the logic for upgrading from the previous minor version.

So to upgrade to v0.8.0, you'd have to download the v0.8.0 osm CLI and run `osm mesh upgrade` which would validate that you're running v0.7.x. If you are indeed running v0.7.x, the upgrade command would contain the logic to remap values if they have changed for v0.8.0.

Not a blocker for this PR. What do you think?

Reply:
Yeah, that approach sounds good to me. I won't plan on adding anything to this PR to handle that, but I'm glad we have a high-level idea of how to approach it.